File: Python FastApi/routers/track_view.py
from fastapi import APIRouter, Depends
from motor.motor_asyncio import AsyncIOMotorDatabase
from typing import List
from datetime import datetime

from dependencies import get_db, public
from models.track_view import TrackViewResponse, MonthlyStats

router = APIRouter()

@router.post("/", response_model=TrackViewResponse)
@public()
async def create_track_view(db: AsyncIOMotorDatabase = Depends(get_db)):
    today = datetime.now().strftime("%Y-%m-%d")
    
    # Check if there's already a record for today
    existing_track = await db.track_views.find_one({"date": today})
    
    if existing_track:
        # Increment view count
        await db.track_views.update_one(
            {"date": today},
            {"$inc": {"views": 1}}
        )
        updated_track = await db.track_views.find_one({"date": today})
        return TrackViewResponse(**updated_track)
    else:
        # Create new record
        track_doc = {
            "date": today,
            "views": 1
        }
        result = await db.track_views.insert_one(track_doc)
        track_doc["_id"] = result.inserted_id
        return TrackViewResponse(**track_doc)

@router.get("/", response_model=List[MonthlyStats])
@public()
async def get_current_year_stats(db: AsyncIOMotorDatabase = Depends(get_db)):
    current_year = datetime.now().year
    
    pipeline = [
        {
            "$match": {
                "date": {"$regex": f"^{current_year}-"}
            }
        },
        {
            "$group": {
                "_id": {"$substr": ["$date", 0, 7]},  # "YYYY-MM"
                "views": {"$sum": "$views"}
            }
        },
        {"$sort": {"_id": 1}}
    ]
    
    result = await db.track_views.aggregate(pipeline).to_list(None)
    
    # Fill in missing months with zero views
    months = []
    for month in range(1, 13):
        month_str = f"{current_year}-{month:02d}"
        month_data = next((item for item in result if item["_id"] == month_str), None)
        months.append({
            "month": month_str,
            "views": month_data["views"] if month_data else 0
        })
    
    return months

@router.get("/month")
@public()
async def get_current_month_total_views(db: AsyncIOMotorDatabase = Depends(get_db)):
    current_month = datetime.now().strftime("%Y-%m")
    
    pipeline = [
        {
            "$match": {
                "date": {"$regex": f"^{current_month}"}
            }
        },
        {
            "$group": {
                "_id": None,
                "totalViews": {"$sum": "$views"}
            }
        }
    ]
    
    result = await db.track_views.aggregate(pipeline).to_list(None)
    
    return result[0]["totalViews"] if result else 0

File: nodejs/routes/track-view.js
const express = require('express');
const TrackView = require('../models/TrackView');

const router = express.Router();

/**
 * @swagger
 * /api/track-view:
 *   post:
 *     summary: Track a view
 *     tags: [Track View]
 *     responses:
 *       200:
 *         description: View tracked successfully
 */
router.post('/', async (req, res, next) => {
  try {
    const today = new Date().toISOString().split('T')[0];

    let track = await TrackView.findOne({ date: today });

    if (track) {
      track.views += 1;
      await track.save();
    } else {
      track = new TrackView({
        date: today,
        views: 1
      });
      await track.save();
    }

    res.json(track);
  } catch (error) {
    next(error);
  }
});

/**
 * @swagger
 * /api/track-view:
 *   get:
 *     summary: Get current year view statistics
 *     tags: [Track View]
 *     responses:
 *       200:
 *         description: Monthly view statistics
 */
router.get('/', async (req, res, next) => {
  try {
    const year = new Date().getFullYear();

    const result = await TrackView.aggregate([
      {
        $match: {
          date: { $regex: `^${year}-` }
        }
      },
      {
        $group: {
          _id: { $substr: ["$date", 0, 7] },
          views: { $sum: "$views" }
        }
      },
      { $sort: { _id: 1 } }
    ]);

    // Fill missing months with zero
    const months = Array.from({ length: 12 }, (_, i) => {
      const month = (i + 1).toString().padStart(2, "0");
      const found = result.find(r => r._id === `${year}-${month}`);
      return {
        month: `${year}-${month}`,
        views: found ? found.views : 0
      };
    });

    res.json(months);
  } catch (error) {
    next(error);
  }
});

/**
 * @swagger
 * /api/track-view/month:
 *   get:
 *     summary: Get current month total views
 *     tags: [Track View]
 *     responses:
 *       200:
 *         description: Current month total views
 */
router.get('/month', async (req, res, next) => {
  try {
    const now = new Date();
    const year = now.getFullYear();
    const month = (now.getMonth() + 1).toString().padStart(2, "0");
    const prefix = `${year}-${month}`;

    const result = await TrackView.aggregate([
      {
        $match: {
          date: { $regex: `^${prefix}` }
        }
      },
      {
        $group: {
          _id: null,
          totalViews: { $sum: "$views" }
        }
      }
    ]);

    const totalViews = result.length > 0 ? result[0].totalViews : 0;
    res.json({ totalViews });
  } catch (error) {
    next(error);
  }
});

module.exports = router;